### DISCUSSION TITLE
posts/elixir-module-registry-via-protocols/

### DISCUSSION THREAD

**Discussion by giscus** [RomanKotov/RomanKotov#12] [Comments]
# posts/elixir-module-registry-via-protocols/

Personal blog

https://romankotov.com/posts/elixir-module-registry-via-protocols.html

**Comment by ribanez7** [comment]
Hi Roman, your article is of great interest to me. At the end, this is the adapter pattern someway. It's curious though that in elixir we lack of a common approach to this pattern without falling into compile time dependencies. For example, let's say we keep the map strategy by doing something like:

```elixir
@models %{
  "model_1" => Module1,
  "model_2" => Module2,
}
```

By doing this, if we have aliased the Module1 and the Module2, we are creating 2 compile time dependencies. If we don't alias them but reference the whole path, we get the same issue. As an alternative: we don't alias them, don't reference the whole module path, and then we do some sort of `Module.concat(whole_path, @models["model_1"])` to solve that compile issue. Another solution could be to have this map in a runtime function. Both these solutions aren't beautiful, the first one is a bad practice, the second one makes us loose any compile time benefit.

That's why I came at your blogpost to read it carefully.

The solution works very well to solve this pattern, but my gotchas:
- The bootstrap maybe is too long and requires a lot of attention at the first time.
- If the model module is not a struct, we are defining a protocol implementation through the macro kind of incorrectly, since defimpl can be only inside a module without `for:` if the module is also a struct. See, for example, your tagged module implementation.

However, I think your approach is the only way to skip compile-time dependencies and keeping an "automatic" approach. Maybe another dummy one would be to create a Module with functions pattern matching the strings "model_1"and "model_2" and dispatching Module1/Module2, but still, it's just too manual.

**Reply by RomanKotov** [reply]
Hi @ribanez7,

Thanks for your feedback!

For me programming is some sort of picking the best alternative :)

Yes, I see your points and agree with them.
- Bootstrap is quite difficult and requires much attention. I think it can be simplified by extracting the logic to a separate library.
- Now Elixir allows `defimpl` for simple modules (not structs), but with limitations. I have tested it on 1.13.3, and it works with extra workarounds.

Regarding the compile-time dependencies. Actually this approach introduces more compile-time dependencies, than a simple map. For example, if we update `ModelExample.Model` all other modules will be recompiled too.


**My Use-case for it:**

I use this approach to gather modules automatically, though it introduces many compile-time dependencies.

For example, it allows me to [lazily generate a mapping of modules for the application](https://github.com/exshome/exshome/blob/a5a0eda6c25f6707a84ee169c2bd4ebb219cb21b/lib/exshome/tag.ex). It returns tag mapping like this:

```elixir
iex(1)> Exshome.Tag.tag_mapping()
%{
  Exshome.Dependency => #MapSet<[Exshome.App.Clock.ClockSettings,
   Exshome.App.Clock.LocalTime, Exshome.App.Clock.UtcTimeService,
   Exshome.App.Player.PlayerService]>,
  Exshome.Named => %{
    "service:player_service" => Exshome.App.Player.PlayerService,
    "service:utc_time_service" => Exshome.App.Clock.UtcTimeService,
    "service_page_live_clock" => ExshomeWeb.Live.ServicePage.ClockPage,
    "service_page_live_player" => ExshomeWeb.Live.ServicePage.PlayerPage,
    "settings:clock_settings" => Exshome.App.Clock.ClockSettings,
    "variable:local_time" => Exshome.App.Clock.LocalTime
  },
  Exshome.Service => #MapSet<[Exshome.App.Clock.UtcTimeService,
   Exshome.App.Player.PlayerService]>,
  Exshome.Settings => #MapSet<[Exshome.App.Clock.ClockSettings]>,
  Exshome.Variable => #MapSet<[Exshome.App.Clock.LocalTime]>,
  ExshomeWeb.Live.ServicePageLive => #MapSet<[ExshomeWeb.Live.ServicePage.ClockPage,
   ExshomeWeb.Live.ServicePage.PlayerPage]>,
  {ExshomeWeb.Live.ServicePageLive, :prefix} => %{
    clock: ExshomeWeb.Live.ServicePage.ClockPage,
    player: ExshomeWeb.Live.ServicePage.PlayerPage
  }
}
```

Then I use this mapping to:
- subscribe and broadcast the changes in any named module through pubsub;
- create generic LiveView pages;
- check that module is registered as a Variable or Service;
- generate routes for pages;
- list all modules, that have the same interface;
- run tests for each similar module/live view page, etc.

It would require many patience from me to keep this mapping up to date. It is OK for me, that any change to `Exshome.Tag` module will require recompilation for depending modules.
